arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -450,114 +458,308 @@ std::unique_ptr<containers::RevisionTree> RocksDBMetaCollection::computeRevision
   return buildTreeFromIterator(it);
 }
 
-bool RocksDBMetaCollection::needToPersistRevisionTree(rocksdb::SequenceNumber maxCommitSeq) const {
-  if (!_logicalCollection.useSyncByRevision()) {
-    return maxCommitSeq > _revisionTreeApplied.load();
+Result RocksDBMetaCollection::takeCareOfRevisionTreePersistence(
+    LogicalCollection& coll, RocksDBEngine& engine,
+    rocksdb::WriteBatch& batch, rocksdb::ColumnFamilyHandle* const cf,
+    rocksdb::SequenceNumber maxCommitSeq,
+    bool force, std::string const& context, std::string& scratch,
+    rocksdb::SequenceNumber& appliedSeq) {
+
+  TRI_ASSERT(coll.useSyncByRevision());
+
+  // might lower `appliedSeq`!
+
+  // Get lock on revision tree:
+  std::unique_lock<std::mutex> guard(_revisionTreeLock);
+  if (maxCommitSeq < _revisionTreeApplied) {
+    // this function is called indirectly from RocksDBSettingsManager::sync which passes
+    // in a seq nr that it has previously obtained, but it can happen that in the meantime
+    // the tree has been moved forward or a tree was rebuilt and _revisionTreeApplied is
+    // now greater than the seq nr we get passed in. In this case we have to bump the seq
+    // nr forward, because we cannot generate a tree from the past.
+    maxCommitSeq = _revisionTreeApplied;
   }
+  
+  maxCommitSeq = _meta.committableSeq(maxCommitSeq);
+  if (needToPersistRevisionTree(maxCommitSeq, guard)) {
```

Comment:
Should also consider force here:
```suggestion
  if (force || needToPersistRevisionTree(maxCommitSeq, guard)) {
```